### PR TITLE
feat(module:upload): start code

### PR DESCRIPTION
**Pull request by thegatheringstorm**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/NG-ZORRO/ng-zorro-antd/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Application (the showcase website) / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #132

## What is the new behavior?

Introduce Upload Component

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```
This code still needs to be perfected in many places.

# Supported API #

- [ ] name

- [ ] defaultFileList

- [x] fileList

- [x] action

- [ ] data

- [x] headers

- [ ] showUploadList

- [x] multiple

- [ ] accept

- [x] beforeUpload

- [ ] ~~customRequest~~

- [x] onChange

- [ ] listType

- [ ] onPreview

- [ ] onRemove

- [ ] ~~supportServerRender~~

- [ ] disabled

- [x] withCredentials

### CODE REVIEW

**Comment by thegatheringstorm** [comment]
## Relationship ##

|NG-ZORRO|Ant-Design|
|-|-|
|[Upload](https://github.com/thegatheringstorm/ng-zorro-antd/tree/master/src/components/upload)|[Upload](https://github.com/ant-design/ant-design/blob/master/components/upload/)|
|[nz-basic-upload](https://github.com/thegatheringstorm/ng-zorro-antd/tree/master/src/components/upload/nz-basic-upload)|[rc-upload](https://github.com/react-component/upload)|

File: package.json [hunk 1]
```diff
@@ -35,23 +35,22 @@
...
+    "@angular/forms": "^4.0.0",
     "@angular/http": "^4.0.0",
     "@angular/language-service": "^4.0.0",
+    "@angular/platform-browser": "^4.0.0",
     "@angular/platform-browser-dynamic": "^4.0.0",
     "@angular/router": "^4.0.0",
     "@types/jasmine": "~2.5.53",
     "@types/jasminewd2": "~2.0.2",
     "@types/node": "~6.0.60",
+    "attr-accept": "^1.1.0",
```
Comment (vthinkxie) [inline package.json]: is it necessary to depend `attr-accept`, there are only 36 lines of code in https://github.com/okonet/attr-accept/blob/master/src/index.js

Comment (thegatheringstorm) [reply package.json]: I think it should be added, sir. It looks clearer.
The 'attr-accept' was also introduced into the Upload component of Ant-Design.
Of course, if you think the inline function is better, I'm pleased to make the change.

Comment (vthinkxie) [reply package.json]: @thegatheringstorm it would better to not depend on unnecessary package since it would bring the risk of environment problem.

Comment (thegatheringstorm) [reply package.json]: @vthinkxie I'm sorry for the late reply.
Sure,there are only 36 lines of code in [attr-accept](https://github.com/okonet/attr-accept/blob/master/src/index.js).But...but it introduce [core-js](https://github.com/zloirock/core-js) in its code.
Should I add [core-js](https://github.com/zloirock/core-js) to my code?

Comment (vthinkxie) [reply package.json]: @thegatheringstorm 
both 'some' and 'endsWith' are supported by typescript.

Comment (thegatheringstorm) [reply package.json]: Thanks!

**Comment by thegatheringstorm** [comment]
I'm sorry that it hasn't been finished for so long. I try to refer to some implementations as much as possible. And complete the task within this week.

**Comment by vthinkxie** [comment]
@thegatheringstorm take it easy, your work is great, and we really appreciate it.

**Comment by thegatheringstorm** [comment]
@vthinkxie I'm so depressed that the CI always fail....
#### 2644
	Error: No provider for HttpClient!
#### 2697
	Can't bind to 'action' since it isn't a known property of 'nz-basic-upload'.
However,it works well on my computer. :sob:

**Comment by gsbybb** [comment]
@thegatheringstorm 会不会是angular版本的问题？ HttpClient 应该在4.3.x才有吧

**Comment by thegatheringstorm** [comment]
@gsbybb Thanks, I will check it later.

**Comment by thegatheringstorm** [comment]
@gsbybb No,I check the version.It fits.....

**Comment by thegatheringstorm** [comment]
Finally,I make basic-upload a component (not module).Fix it.

**Comment by thegatheringstorm** [comment]
The PR will be reopened after all tests pass soon.

**Pull request closed without merging**